python/tempo/tsdf.py

```diff
@@ -34,15 +34,38 @@ def __init__(self, df, ts_col="event_ts", partition_cols=None, sequence_col = No
 
     self.df = df
     self.sequence_col = '' if sequence_col is None else sequence_col
+
+    ## Add customized check for string type for the timestamp. If we see a string, we will proactively created a double version of the string timestamp for sorting purposes and rename to ts_col
+    if (df.schema[ts_col].dataType == "StringType"):
+        sample_ts = df.limit(1).collect()[0][0]
+        self.__validate_ts_string(sample_ts)
+        self.__add_double_ts().withColumnRenamed("double_ts", self.ts_col)
+
     """
     Make sure DF is ordered by its respective ts_col and partition columns.
     """
 
+
   ##
   ## Helper functions
   ##
 
-  def __validated_column(self,df,colname):
+  def __add_double_ts(self):
+      """ Add a double (epoch) version of the string timestamp out to nanos
+      """
+      self.df = self.df.withColumn("nanos", (f.when(f.col(self.ts_col).contains("."),
+                                     f.concat(f.lit("0."), f.split(f.col(self.ts_col), '\.')[1])).otherwise(0)).cast(
+          "double")).withColumn("double_ts", f.col(self.ts_col).cast("long") + f.col("nanos"))
+
+  def __validate_ts_string(self, ts_text):
+      """ Validate the format for the string"""
+      import datetime
+      try:
+          datetime.datetime.strptime(ts_text, '%Y-%m-%dT%H:%M:%S')
```

Comment:
I don't believe this format will accept a timestamp with fractional seconds. When I tested this pattern against a timestamp with fractional seconds, I got a ValueError. I'm not sure that the `strptime` function can accept nanos - the docs only talk about the %f pattern directive which supports microseconds, but not nanos (it supports up to 6 digits of fractional seconds, no more). 

I think we may need a custom REGEX here.  Try this:
```
import re
ts_pattern = "^\d{4}-\d{2}-\d{2}T| \d{2}:\d{2}:\d{2}\.\d*$"
if re.match(ts_pattern, ts_text) is None:
  raise ValueError("Incorrect data format, should be YYYY-MM-DD HH:MM:SS[.nnnnnnnn]")
```

Reply:
Good catch, this was tested out to seconds. It has been addressed with a new update.

Reply:
Looks great

Reply:
Actually we need the 'T' in that expression - for ISO formatted timestamps. I'm adding it back in...